server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -57,15 +102,24 @@ public String toString() {
 
     @Override
     public void allocate(RoutingAllocation allocation) {
+        allocate(allocation, ActionListener.noop());
+    }
+
+    @Override
+    public void allocate(RoutingAllocation allocation, ActionListener<Void> listener) {
         assert MasterService.isMasterUpdateThread() || Thread.currentThread().getName().startsWith("TEST-")
             : Thread.currentThread().getName();
         // assert allocation.debugDecision() == false; set to true when called via the reroute API
         assert allocation.ignoreDisable() == false;
 
         // TODO must also capture any shards that the existing-shards allocators have allocated this pass, not just the ignored ones
 
+        var index = indexGenerator.incrementAndGet();
+        synchronized (pendingListeners) {
+            pendingListeners.addLast(new DesiredBalancesListener(index, listener));
```

Comment:
Do we need to preserve the thread context of this listener? That's generally a problem with collecting listeners and then completing them all later. But I guess we're in system context here and we also complete them all in system context so maybe not. We should be able to assert that we're in system context tho (with some test adjustments because the tests won't be running this in system context I think)